### PR TITLE
[Feature] lake tablet support auto partition

### PR DESCRIPTION
**Pull request by kevincai**
## What type of PR is this：
- [ ] BugFix
- [X] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [X] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 9 / 18 (50.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|9|18|50.00%|[1837, 1838, 1840, 1841, 1850, 1852, 1853, 1854, 1855]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 9 / 18 (50.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|9|18|50.00%|[1837, 1838, 1840, 1841, 1850, 1852, 1853, 1854, 1855]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 9 / 18 (50.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|9|18|50.00%|[1837, 1838, 1840, 1841, 1850, 1852, 1853, 1854, 1855]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 9 / 19 (47.37%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|9|19|47.37%|[1770, 1839, 1840, 1842, 1843, 1852, 1854, 1855, 1856, 1857]|

**Review by meegoo — CHANGES_REQUESTED** [review]
LakeTabletsChannel::open need update too

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 9 / 19 (47.37%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|9|19|47.37%|[1770, 1839, 1840, 1842, 1843, 1852, 1854, 1855, 1856, 1857]|

**Comment by kevincai** [comment]
> LakeTabletsChannel::open need update too

done. please take a look again.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 9 / 19 (47.37%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|9|19|47.37%|[1813, 1882, 1883, 1885, 1886, 1895, 1897, 1898, 1899, 1900]|

File: be/src/runtime/load_channel.cpp [hunk 1]
```diff
@@ -142,7 +142,7 @@ void LoadChannel::open(brpc::Controller* cntl, const PTabletWriterOpenRequest& r
                     st = local_tablets_channel->incremental_open(request, _schema);
                 }
             } else {
-                st = Status::NotSupported("incremental open not supported by this tablets channel");
+                st = it->second->incremental_open(request, _schema);
```
Comment (meegoo) [inline be/src/runtime/load_channel.cpp:145]: there is concurrent issue between write and incremental open?

Comment (kevincai) [reply be/src/runtime/load_channel.cpp:123]: done

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 9 / 25 (36.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|9|25|36.00%|[1819, 1888, 1889, 1890, 1891, 1892, 1894, 1895, 1896, 1898, 1899, 1908, 1910, 1911, 1912, 1913]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 9 / 25 (36.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|9|25|36.00%|[1819, 1888, 1889, 1890, 1891, 1892, 1894, 1895, 1896, 1898, 1899, 1908, 1910, 1911, 1912, 1913]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 9 / 25 (36.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|9|25|36.00%|[1819, 1888, 1889, 1890, 1891, 1892, 1894, 1895, 1896, 1898, 1899, 1908, 1910, 1911, 1912, 1913]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 9 / 25 (36.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|9|25|36.00%|[1819, 1888, 1889, 1890, 1891, 1892, 1894, 1895, 1896, 1898, 1899, 1908, 1910, 1911, 1912, 1913]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 9 / 25 (36.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|9|25|36.00%|[1819, 1888, 1889, 1890, 1891, 1892, 1894, 1895, 1896, 1898, 1899, 1908, 1910, 1911, 1912, 1913]|

**Review by starrocks-xupeng — APPROVED** [review]

File: be/src/util/await.h [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+#include <functional>
+
+namespace starrocks {
+
+// Convenient function to archive wait-until ability
+// Awaitility()
+// .timeout(1000 * 1000) // one second
+// .interval(1000)  // check every 1000 microseconds
+// .until(() -> {}); // until the condition is true or timedout
+class Awaitility {
```
Comment (sduzh) [inline be/src/util/await.h:26]: This class addresses this need elegantly!

**Review by tracymacding — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/service/FrontendServiceImpl.java [hunk 3]
```diff
@@ -1879,25 +1884,40 @@ private static TCreatePartitionResult createPartitionProcess(TCreatePartitionReq
...
-                        errorStatus.setError_msgs(Lists.newArrayList(
-                                "Tablet lost replicas. Check if any backend is down or not. tablet_id: "
-                                        + tablet.getId() + ", backends: " +
-                                        Joiner.on(",").join(localTablet.getBackends())));
-                        result.setStatus(errorStatus);
-                        return result;
+                if (olapTable.isCloudNativeTable()) {
+                    for (Tablet tablet : index.getTablets()) {
+                        LakeTablet cloudNativeTablet = (LakeTablet) tablet;
+                        Set<Long> backendIds = cloudNativeTablet.getBackendIds();
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/service/FrontendServiceImpl.java:1890]: only need primary backend

File: be/src/runtime/tablets_channel.cpp [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+namespace starrocks {
+
+bool TabletsChannel::drain_senders(int64_t timeout, const std::string& log_msg) {
+    int64_t check_interval = 10 * 1000;       // 10ms
+    auto duration = std::chrono::seconds(60); // 60 seconds;
+    auto start = std::chrono::steady_clock::now();
+    auto next_logging = start + duration;
+
+    auto cb = [&](int i) {
```
Comment (wyb) [inline be/src/runtime/tablets_channel.cpp:28]: `i` is not used

File: be/test/util/bthreads/shared_mutex_test.cpp [hunk 5]
```diff
@@ -0,0 +1,182 @@
+// Copyright 2023-present StarRocks, Inc. All rights reserved.
```
[the remark this answers is not in the document]
Comment (kevincai) [reply be/test/util/bthreads/shared_mutex_test.cpp]: done

File: fe/fe-core/src/main/java/com/starrocks/service/FrontendServiceImpl.java (hunk 3 above)
Comment (kevincai) [reply fe/fe-core/src/main/java/com/starrocks/service/FrontendServiceImpl.java:1890]: local tablet provides all the locations, any difference here?

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 9 / 25 (36.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|9|25|36.00%|[1819, 1888, 1889, 1890, 1891, 1892, 1894, 1895, 1896, 1898, 1899, 1908, 1910, 1911, 1912, 1913]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 9 / 24 (37.50%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|9|24|37.50%|[1792, 1793, 1794, 1795, 1796, 1798, 1799, 1800, 1802, 1803, 1812, 1814, 1815, 1816, 1817]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 9 / 24 (37.50%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|9|24|37.50%|[1792, 1793, 1794, 1795, 1796, 1798, 1799, 1800, 1802, 1803, 1812, 1814, 1815, 1816, 1817]|

**Review by meegoo — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/service/FrontendServiceImpl.java (hunk 3 above)
Comment (kevincai) [reply fe/fe-core/src/main/java/com/starrocks/service/FrontendServiceImpl.java]: done

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 9 / 25 (36.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|9|25|36.00%|[1792, 1793, 1796, 1797, 1798, 1799, 1801, 1802, 1803, 1804, 1805, 1814, 1816, 1817, 1818, 1819]|

**Review by meegoo — APPROVED** [review]

**Review by wyb — APPROVED** [review]

**Pull request merged**